docs/running_lbann.rst

```diff
@@ -1,96 +1,435 @@
+    (objective function gradients w.r.t. input tensors) to its
+    parents. If the layer has any associated weight tensors, it will
+    also compute objective function gradients w.r.t. the weight
+    tensors.
+
+  - Most layers require a specific number of parents and children, but
+    LBANN will insert layers into the graph if there is a mismatch and
+    the intention is obvious. For example, if a layer expects one
+    child but has multiple, then a split layer (with multiple output
+    tensors all identical to the input tensor) is inserted. Similarly,
+    if a layer has fewer children than expected, dummy layers will be
+    inserted. However, this does not work if there is any
+    ambiguity. In such cases (common with input and slice layers), it
+    is recommended to manually insert identity layers so that the
+    parent/child relationships are absolutely unambiguous.
+
++ Weights: A tensor consisting of trainable parameters, typically
+  associated with one or more layers. A weight tensor owns an
+  initializer to initially populate its values and an optimizer to
+  find values that minimize the objective function.
+
+  - A weight tensor without a specified initializer will use a zero
+    initializer.
+
+  - A weight tensor without a specified optimizer will use the model's
+    default optimizer.
+
+  - If a layer requires weight tensors and none are specified, it will
+    create the needed weight tensors. The layer will pick sensible
+    initializers and optimizers for the weight tensors.
+
+  - The dimensions of a weight tensor is determined by their
+    associated layers. The user can not set it directly.
+
++ Objective function: Mathematical expression that the optimizers will
+  attempt to minimize. It is made up of multiple terms that are added
+  together (possibly with scaling factors).
+
+  - An objective function term can get its value from a scalar-valued
+    layer, i.e. a layer with an output tensor with one entry.
+
++ Metric: Mathematical expression that will be reported to the
+  user. This typically does not affect training, but is helpful for
+  evaluating the progress of training. A canonical example for
+  classification problems is classification accuracy.
+
++ Callback: Function that is performed at various points during an
+  experiment. Callbacks are helpful for reporting, debugging, and
+  performing advanced training techniques.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Data readers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning:: The core infrastructure for data readers is slated for
+             significant refactoring, so expect major changes in the
+             future.
+
+Data readers are responsible for managing a data set and providing
+data samples to models. A data set is comprised of independent data
+samples, each of which is made up of multiple tensors. For example, a
+data sample for a labeled image classification problem consists of an
+image tensor and a one-hot label vector.
```

Comment:
May be worthwhile to mention that label vector is _N/A_ for unsupervised learning

Reply:
Labels are only specific to classification problems. For unsupervised learning, the data reader should only output one tensor (once we get around to properly refactoring the data reader).

Reply:
Yes, but right now the user would have to specify _target_mode_ for input layer to work. I understand is just an example, may be it is ok not cover all _possible modes_.